### PR TITLE
Add IndicateVarDataType interface to block tensor is not initialized problem in OP GetExceptedKernelType

### PR DESCRIPTION
**Pull request by chenwhql**
In some cases, paddle will throw errors like `holder_ should not be null. Tensor not initialized yet when Tensor::type() is called.`. This error is caused by the direct use of the tensor in the `Variable` without initialization, but this error message is not user friendly.

By analyzing issues, I found that `GetExceptedKernelType` is an important source of this problem. Our developers get the tensor type of input variable but don't check if the tensor was initialized. So I think we can add a funciton to deal with this problem.

If Op doesn't overwrite the `GetExceptedKernelType` function, use the `OperatorWithKernel::GetExpectedKernelType` function, it wil check if tensor was initialized, so I add the `IndicateVarDataType` which is just did what we should have done.

Then the interface `GetDataTypeOfVar` can be removed, and we should add a rule that developers must use this `IndicateVarDataType` interface in Ops' `GetExceptedKernelType`.

### The error cases:

https://github.com/PaddlePaddle/Paddle/issues/19522

image

https://github.com/PaddlePaddle/Paddle/issues/18432

image

### This PR's results:

- original error message:

```
holder_ should not be null
Tensor not initialized yet when Tensor::type() is called. at [/paddle/paddle/fluid/framework/tensor.h:139]
```

- new error message:

```
The Tensor in the save Op's Input Variable X(mean_0.tmp_0) is not initialized. at [/work/paddle/paddle/fluid/framework/operator.cc:1162]
  [operator < save > error]
```

## Use Cases

1. `GetDataTypeOfVar(ctx.InputVar("X")) -> IndicateVarDataType(ctx, "X")`

2. `ctx.Input<Tensor>("X")->type() -> IndicateVarDataType(ctx, "X")`

### Code
```
import paddle.fluid as fluid

place = fluid.CPUPlace()
exe = fluid.Executor(place)

x = fluid.layers.data(name='X', shape=[13], dtype='float32')
y = fluid.layers.data(name='Y', shape=[1], dtype='float32')
y_ = fluid.layers.fc(input=x, size=1, act=None)
loss = fluid.layers.square_error_cost(input=y_, label=y)
avg_loss = fluid.layers.mean(loss)
avg_loss.persistable = True

fluid.optimizer.SGD(learning_rate=0.01).minimize(avg_loss)

exe.run(fluid.default_startup_program())
fluid.io.save_inference_model(dirname="./save_op_test",
                              feeded_var_names=['X', 'Y'],
                              target_vars=[avg_loss],
                              executor=exe)
```

### Result

```
--------------------------------------------
C++ Call Stacks (More useful to developers):
--------------------------------------------
0   std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > paddle::platform::GetTraceBackString<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*, int)
1   paddle::platform::EnforceNotMet::EnforceNotMet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*, int)
2   paddle::framework::OperatorWithKernel::ParseInputDataType(paddle::framework::ExecutionContext const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, paddle::framework::proto::VarType_Type*) const
3   paddle::framework::OperatorWithKernel::IndicateVarDataType(paddle::framework::ExecutionContext const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
4   paddle::operators::SaveOp::GetExpectedKernelType(paddle::framework::ExecutionContext const&) const
5   paddle::framework::OperatorWithKernel::ChooseKernel(paddle::framework::RuntimeContext const&, paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const
6   paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&, paddle::framework::RuntimeContext*) const
7   paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const
8   paddle::framework::OperatorBase::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&)
9   paddle::framework::Executor::RunPreparedContext(paddle::framework::ExecutorPrepareContext*, paddle::framework::Scope*, bool, bool, bool)
10  paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool)

------------------------------------------
Python Call Stacks (More useful to users):
------------------------------------------
  File "/usr/local/lib/python3.5/dist-packages/paddle/fluid/framework.py", line 2423, in append_op
    attrs=kwargs.get("attrs", None))
  File "/usr/local/lib/python3.5/dist-packages/paddle/fluid/io.py", line 241, in save_vars
    attrs={'file_path': save_file_path})
  File "/usr/local/lib/python3.5/dist-packages/paddle/fluid/io.py", line 216, in save_vars
    filename=filename)
  File "/usr/local/lib/python3.5/dist-packages/paddle/fluid/io.py", line 571, in save_persistables
    filename=filename)
  File "/usr/local/lib/python3.5/dist-packages/paddle/fluid/io.py", line 1197, in save_inference_model
    save_persistables(executor, save_dirname, main_program, params_filename)
  File "save_op.py", line 22, in <module>
    executor=exe)

----------------------
Error Message Summary:
----------------------
PaddleCheckError: Expected t->IsInitialized() == true, but received t->IsInitialized():0 != true:1.
The Tensor in the save Op's Input Variable X(mean_0.tmp_0) is not initialized. at [/work/paddle/paddle/fluid/framework/operator.cc:1162]
  [operator < save > error]
```

### CODE REVIEW

File: paddle/fluid/framework/operator.cc [hunk 1]
```diff
@@ -1175,6 +1175,34 @@ proto::VarType::Type OperatorWithKernel::IndicateDataType(
...
+  const Tensor* t = nullptr;
+  if (var->IsType<LoDTensor>()) {
+    t = &var->Get<LoDTensor>();
+  } else if (var->IsType<SelectedRows>()) {
+    t = &(var->Get<SelectedRows>().value());
+  } else {
+    PADDLE_THROW(
+        "The Input Variable(%s) of %s Op used to determine kernel data type "
+        "should be LoDTensor or SelectedRows.",
+        name, this->type_);
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc]: The Variable's name should be `ctx->Inputs().frond()`.

File: paddle/fluid/framework/operator.cc [hunk 2]
```diff
@@ -1175,6 +1175,34 @@ proto::VarType::Type OperatorWithKernel::IndicateDataType(
...
+  }
+  // No need to check if t is nullptr, if so, var->isType or var->Get will throw
+  // error.
+  PADDLE_ENFORCE_EQ(
+      t->IsInitialized(), true,
+      "The Tensor in the %s Op's Input Variable %s is not initialized.",
+      this->type_, name);
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc]: Same above.

File: paddle/fluid/framework/operator.cc [hunk 3]
```diff
@@ -1175,6 +1175,34 @@ proto::VarType::Type OperatorWithKernel::IndicateDataType(
...
+  return t->type();
+}
+
```
Comment (chengduoZH) [inline paddle/fluid/framework/operator.cc:1202]: Some logic of `OperatorWithKernel::IndicateDataType(const ExecutionContext& ctx, const std::string& name)` and `OperatorWithKernel::IndicateDataType(const ExecutionContext& ctx)` are the same, maybe this part can be shared.

File: paddle/fluid/framework/operator.cc [hunk 4]
```diff
@@ -1175,6 +1175,34 @@ proto::VarType::Type OperatorWithKernel::IndicateDataType(
...
 }
 
+proto::VarType::Type OperatorWithKernel::IndicateVarDataType(
+    const ExecutionContext& ctx, const std::string& name) const {
+  const std::vector<const Variable*> vars = ctx.MultiInputVar(name);
+  PADDLE_ENFORCE_EQ(
+      vars.empty(), false,
+      "The %s Op doesn't have Input %s or it is an empty Variable.",
+      this->type_, name);
+  const Variable* var = vars.front();
```
Comment (sneaxiy) [inline paddle/fluid/framework/operator.cc]: Why get `front()` only?

File: paddle/fluid/framework/variable.h [hunk 5]
```diff
@@ -27,12 +27,12 @@ class Variable {
  public:
   template <typename T>
   const T& Get() const {
-    static_assert(
-        IsRegisteredVarType<T>(),
-        "Not registered type. Please register T inside var_type_traits.h");
-    PADDLE_ENFORCE(holder_ != nullptr, "Variable must hold some thing");
+    static_assert(IsRegisteredVarType<T>(),
```
Comment (sneaxiy) [inline paddle/fluid/framework/variable.h]: `static_assert` only throws to users when compiling PaddlePaddle. It is not necessary to change this line.

File: paddle/fluid/framework/variable.h [hunk 6]
```diff
@@ -27,12 +27,12 @@ class Variable {
...
+                  "Type T is not registered. Please register T inside "
+                  "var_type_traits.h.");
+    PADDLE_ENFORCE(holder_ != nullptr, "This Variable is not initialized.");
```
Comment (sneaxiy) [inline paddle/fluid/framework/variable.h]: `This Variable` is also confused to users. Change it to `Variable`.

File: paddle/fluid/framework/variable.h [hunk 7]
```diff
@@ -87,9 +87,9 @@ class Variable {
   // parameter of Variable.
   template <typename T>
   struct PlaceholderImpl : public Placeholder {
-    static_assert(
-        IsRegisteredVarType<T>(),
-        "Not registered type. Please register T inside var_type_traits.h");
+    static_assert(IsRegisteredVarType<T>(),
```
Comment (sneaxiy) [inline paddle/fluid/framework/variable.h]: `static_assert` only throws to users when compiling PaddlePaddle. It is not necessary to change this line.

File: paddle/fluid/framework/operator.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/framework/operator.cc]: The `ctx->Inputs().front()` maybe the generated name, I think use the proto name is better, and I have changed to check all vars in the specific input, if error occured on the var after the first var, we use `front()` at here is not accurate·

File: paddle/fluid/framework/operator.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/framework/operator.cc]: we can use both of them in this place, like `Variable X(fc_0.tmp_1)`

File: paddle/fluid/framework/operator.cc (hunk 3 above)
Comment (chenwhql) [reply paddle/fluid/framework/operator.cc:1202]: agree, I try to extract public parts of two functions.

File: paddle/fluid/framework/operator.cc (hunk 4 above)
Comment (chenwhql) [reply paddle/fluid/framework/operator.cc]: currently, the `MultiInputVar` used in `GetExpectedKernelType`only has following two forms:
`ctx.MultiInputVar("X")[0]`
`ctx.MultiInputVar("Ids").front()`
So I only get `front()`, but it is hard, already fixed

File: paddle/fluid/framework/variable.h (hunk 5 above)
Comment (chenwhql) [reply paddle/fluid/framework/variable.h]: Done

File: paddle/fluid/framework/variable.h (hunk 7 above)
Comment (chenwhql) [reply paddle/fluid/framework/variable.h]: Done

File: paddle/fluid/framework/variable.h (hunk 6 above)
Comment (chenwhql) [reply paddle/fluid/framework/variable.h]: Done

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Review by luotao1 — APPROVED** [review]
LGTM

**Pull request merged**